src/core/Akka.Cluster/Cluster.cs

```diff
@@ -264,27 +267,24 @@ public void Join(Address address)
         /// <returns>Task which completes, once current cluster node reaches <see cref="MemberStatus.Up"/> state.</returns>
         public Task JoinAsync(Address address, CancellationToken token = default)
         {
-            var completion = new TaskCompletionSource<NotUsed>(TaskCreationOptions.RunContinuationsAsynchronously);
-            
-            var timeout = Settings.RetryUnsuccessfulJoinAfter ?? TimeSpan.FromSeconds(10);
-            var timeoutCts = CancellationTokenSource.CreateLinkedTokenSource(token);
-            timeoutCts.CancelAfter(timeout);
-            timeoutCts.Token.Register(() =>
+            if (_asyncJoinState != null)
```

Comment:
This could still fail under a number of scenarios including two different threads calling JoinAsync 'close enough'

Given this is not a hot path, we probably could get away with a quick `lock` here; An uncontended lock will be fast (Which is the happy path) and is only needed for the time it takes to check the state and new up the Task/AsyncState